stackslib/src/net/download/nakamoto.rs

```diff
@@ -0,0 +1,4283 @@
+    /// Returns the tenures from first_block_height (inclusive) to last_block_height (exclusive) on
+    /// success.
+    ///
+    /// Returns Err(..) on DB error, or if one or both of these heights do not correspond to a
+    /// sortition.
+    pub(crate) fn load_wanted_tenures(
+        ih: &SortitionHandleConn,
+        first_block_height: u64,
+        last_block_height: u64,
+    ) -> Result<Vec<WantedTenure>, NetError> {
+        let mut wanted_tenures = Vec::with_capacity(
+            usize::try_from(last_block_height.saturating_sub(first_block_height))
+                .expect("FATAL: infallible: usize can't old a reward cycle"),
+        );
+        let mut cursor = ih
+            .get_block_snapshot_by_height(last_block_height.saturating_sub(1))?
+            .ok_or(DBError::NotFoundError)?;
+        while cursor.block_height >= first_block_height {
+            test_debug!(
+                "Load sortition {}/{} burn height {}",
+                &cursor.consensus_hash,
+                &cursor.winning_stacks_block_hash,
+                cursor.block_height
+            );
+            wanted_tenures.push(WantedTenure::new(
+                cursor.consensus_hash,
+                StacksBlockId(cursor.winning_stacks_block_hash.0),
+                cursor.block_height,
+            ));
+            cursor = SortitionDB::get_block_snapshot(&ih, &cursor.parent_sortition_id)?
+                .ok_or(DBError::NotFoundError)?;
+        }
+        wanted_tenures.reverse();
+        Ok(wanted_tenures)
+    }
+
+    /// Find the list of wanted tenures for the given reward cycle.  The reward cycle must
+    /// be complete already.  Used for testing.
+    ///
+    /// Returns a reward cycle's wanted tenures.
+    /// Returns a DB error if the snapshot does not correspond to a full reward cycle.
+    #[cfg(test)]
+    pub(crate) fn load_wanted_tenures_for_reward_cycle(
+        cur_rc: u64,
+        tip: &BlockSnapshot,
+        sortdb: &SortitionDB,
+    ) -> Result<Vec<WantedTenure>, NetError> {
+        // careful -- need .saturating_sub(1) since this calculation puts the reward cycle start at
+        // block height 1 mod reward cycle len, but we really want 0 mod reward cycle len
+        let first_block_height = sortdb
+            .pox_constants
+            .reward_cycle_to_block_height(sortdb.first_block_height, cur_rc)
+            .saturating_sub(1);
+        let last_block_height = sortdb
+            .pox_constants
+            .reward_cycle_to_block_height(sortdb.first_block_height, cur_rc.saturating_add(1))
+            .saturating_sub(1);
+
+        test_debug!(
+            "Load reward cycle sortitions between {} and {} (rc is {})",
+            first_block_height,
+            last_block_height,
+            cur_rc
+        );
+
+        // find all sortitions in this reward cycle
+        let ih = sortdb.index_handle(&tip.sortition_id);
+        Self::load_wanted_tenures(&ih, first_block_height, last_block_height)
+    }
```

Comment:
Test specific code should live in test modules unless absolutely necessary.

Reply:
Sure